mlir-sycl/include/mlir/Dialect/SYCL/Analysis/ConstructorBaseAnalysis.h

```diff
@@ -0,0 +1,158 @@
+//===- ConstructorBaseAnalysis.h - Base analysis for SYCL types -*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Base class for analyses that determine properties of instances of SYCL types
+// based on their construction.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_DIALECT_SYCL_ANALYSIS_CONSTRUCTORBASEANALYSIS_H
+#define MLIR_DIALECT_SYCL_ANALYSIS_CONSTRUCTORBASEANALYSIS_H
+
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/Polygeist/Analysis/DataFlowSolverWrapper.h"
+#include "mlir/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.h"
+#include "mlir/Dialect/SYCL/Analysis/AliasAnalysis.h"
+#include "mlir/Pass/AnalysisManager.h"
+
+namespace mlir {
+namespace sycl {
+
+/// CRTP base class for analyses deriving properties of interest for SYCL
+/// objects from their construction. The analyses derived from this class, i.e.,
+/// `ConcreteAnalysis` must implement the following methods:
+/// * `void finalizeInitialization(bool useRelaxedAliasing)`
+/// * `bool isConstructorImpl(const polygeist::Definition &def)`
+/// * `AnalysisResult getInformationImpl(const polygeist::Definition &def)`
+///
+/// The class representing information derived by the analysis, i.e.,
+/// `AnalysisResult` must define a default constructor and must implement the
+/// following methods:
+/// * `bool isTop()` to indicate the top of the lattice has been reached.
+/// * `AnalysisResult join(const AnalysisResult& other)` to join two results at
+/// a merge point.
+///
+template <typename ConcreteAnalysis, typename AnalysisResult>
+class ConstructorBaseAnalysis {
```

Comment:
Should it be `ConstructorBasedAnalysis`?

Reply:
Well, it's the base class for constructor-based analysis, so I'm not sure there's a right or wrong here 😄

Reply:
Yes, we can leave this as is :laughing: